### ISSUE TITLE
new rules for Ansible 2.4 import_tasks

### ISSUE THREAD

**Issue by wookietreiber** [ansible/ansible-lint#306]
# Issue Type
- Feature request

# Desired Behaviour

Ansible 2.4 introduced the new `import_tasks` (static include) and `include_tasks` (dynamic include). Since `import_tasks` should only be used for completely static things, I think, it would be a good idea to issue a warning on the following two conditions:

1.  `import_tasks` is used with a `when`, e.g.:

    ```yml
    - import_tasks: foo.yml
      when: foo_condition
    ```

    The inclusion of `foo.yml` is *inherently dynamic* because it depends on the condition `foo_condition`.

1.  `import_tasks` is used with a variable file, e.g.:

    ```yml
    - import_tasks: 'build-deps-{{ ansible_os_family }}.yml'
    ```

    The choice which file is included is *inherently dynamic*.

What is your opinion on these rules?

**Comment by techzilla** [comment]
We definitely need additional rules for the 2.4 changes,

1. deprecated use of include:
2. incorrect use of import_ when include_ was required.

**Comment by techzilla** [comment]
Not sure if this is possible with the present code base structure, you can't really get at the `includes:`.

**Comment by mikef-nl** [comment]
[IncludeDeprecated.zip](https://github.com/willthames/ansible-lint/files/1879488/IncludeDeprecated.zip)
@techzilla well that's 1. down

**Comment by mjrider** [comment]
in http://docs.ansible.com/ansible/latest/modules/import_tasks_module.html
we have the following example:
```
- hosts: all
  tasks:
    - name: Apply conditional to all imported tasks
      import_tasks: stuff.yaml
      when: hostvar is defined
```

So the ansible documentation seem to suggest that the first case, using when with import_tasks is fine

**Comment by wookietreiber** [comment]
From the original 2.4 [changelog](https://github.com/ansible/ansible/blob/v2.4.0.0-1/CHANGELOG.md):

- `import_*` are static
- `include_*` are dynamic

This implies, that:

1.  `include_*` SHOULD be used if e.g. the file name contains variables; variables make tasks dynamic
1.  `include_*` SHOULD be used if there is a conditional; conditionals make tasks dynamic
1.  `import_*` SHOULD be used only if everything is static

---

I use e.g. `import_tasks` only when I split up a role in multiple files for the purpose of separation of concerns, e.g. when a role installs some cron jobs, I want them to be in a `cron.yml` which I import from the `main.yml` to keep everything neat and clean.

**Comment by wookietreiber** [comment]
Unfortunately, the new [porting guide](http://docs.ansible.com/ansible/latest/porting_guides/porting_guide_2.4.html) does not contain information about it.

**Comment by k0ste** [comment]
At this point I porting roles to 2.5, so after post from @wookietreiber this is simple:

```yaml
---
- include_tasks: 'packages.yml' # condition = include
  when:
  - "hostvars[inventory_hostname]['install_package'] is defined"
  - "hostvars[inventory_hostname]['install_package'] == 'true'"
- import_tasks: 'deploy_something.yml' # no condition = import
```

**Comment by wilmardo** [comment]
```yaml
- import_tasks: 'build-deps-{{ ansible_os_family }}.yml'
```

This is valid tho, you can template an import from vars (even from defaults atm but it always takes the role default never any override).

In my opinion every plain include should be static:

```yaml
- include_tasks: packages.yml
```

And every import with a when is just wrong:

```yaml
- import_tasks: packages.yml
  when: test is defined
```

**Comment by ssbarnea** [comment]
Please reopen if you can reproduce bug with master code and include the file.
